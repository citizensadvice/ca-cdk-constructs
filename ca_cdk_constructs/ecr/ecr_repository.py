from aws_cdk.aws_ecr import Repository, LifecycleRule, TagStatus
from aws_cdk import Duration, aws_iam as iam
from constructs import Construct


class ECRRepository(Construct):
    """
    An ECR repository with default lifecycle rules.

    The default rules are as follows:
    - Remove all untagged images after 1 day
    - Remove all development images after 90 days. Dev image tags start with `dev_`
    - Limit the max number of images to 1000

    Max development images and max images are configurable.

    :param scope: The scope of the construct, usually self.
    :param id: The id of the construct.
    :param name: The name of the repository. Must be unique within the account.
    :param additional_lifecycle_rules: Additional lifecycle rules to add in addition to the default rules. Priority must be unique and equal to or less than 10 as not to conflict with existing rules.
    :param additional_accounts_pull: Additional accounts that are allowed to pull from the repository in the form of their account number
    :param additional_accounts_push: Additional accounts that are allowed to push to the repository in the form of their account number.
    :param dev_image_max_age: The maximum age of a development image in days. Defaults to 90.
    :param max_images: The maximum number of images allowed in the repository. Defaults to 1000.
    :param scan_on_push: Whether to scan for vulnerabilities on the image on push. Defaults to True.
    """

    def __init__(
        self,
        scope: Construct,
        id: str,
        name: str,
        additional_lifecycle_rules: list[LifecycleRule] = [],
        additional_accounts_pull: list[str] = [],
        additional_accounts_push: list[str] = [],
        dev_image_max_age: int = 90,
        max_images: int = 1000,
        scan_on_push: bool = True,
    ):
        super().__init__(scope, id)
        lifecycle_rules = []

        lifecycle_rules.append(
            LifecycleRule(
                rule_priority=20,
                description="Limit max number of images",
                max_image_count=max_images,
                tag_status=TagStatus.ANY,
            )
        )
        lifecycle_rules.append(
            LifecycleRule(
                rule_priority=19,
                description="Delete untagged images",
                tag_status=TagStatus.UNTAGGED,
                max_image_age=Duration.days(1),
            )
        )
        lifecycle_rules.append(
            LifecycleRule(
                rule_priority=18,
                description="Delete development images after 90 days year",
                tag_status=TagStatus.TAGGED,
                tag_prefix_list=["dev_"],
                max_image_age=Duration.days(dev_image_max_age),
            )
        )

        lifecycle_rules += additional_lifecycle_rules

        self.repository = Repository(
            self,
            "Default",
            repository_name=name,
            lifecycle_rules=lifecycle_rules,
            image_scan_on_push=scan_on_push,
        )

        for account in additional_accounts_pull:
            self.repository.grant_pull(iam.AccountPrincipal(account))

        for account in additional_accounts_push:
            self.repository.grant_push(iam.AccountPrincipal(account))
