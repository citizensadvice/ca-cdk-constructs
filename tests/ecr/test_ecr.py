from ca_cdk_constructs.ecr.ecr_repository import ECRRepository
from aws_cdk import Stack, App, assertions
from aws_cdk.aws_ecr import LifecycleRule, TagStatus
import pytest


@pytest.fixture(scope="function")
def stack():
    return Stack(App(), "TestStack")


def test_ecr_defaults(stack, snapshot):
    ECRRepository(stack, "TestECRRepository", name="test-repository")
    assert assertions.Template.from_stack(stack).to_json() == snapshot


def test_ecr_additional_rules(stack, snapshot):
    ECRRepository(
        stack,
        "TestECRRepository",
        name="test-repository",
        additional_lifecycle_rules=[
            LifecycleRule(
                description="Test rule",
                rule_priority=10,
                max_image_count=10,
                tag_status=TagStatus.UNTAGGED,
            )
        ],
    )
    assert assertions.Template.from_stack(stack).to_json() == snapshot


def test_ecr_additional_accounts_pull(stack, snapshot):
    ECRRepository(
        stack,
        "TestECRRepository",
        name="test-repository",
        additional_accounts_pull=["123456789012"],
    )
    assert assertions.Template.from_stack(stack).to_json() == snapshot


def test_ecr_additional_accounts_push(stack, snapshot):
    ECRRepository(
        stack,
        "TestECRRepository",
        name="test-repository",
        additional_accounts_push=["123456789012"],
    )
    assert assertions.Template.from_stack(stack).to_json() == snapshot


def test_ecr_scan_disabled(stack, snapshot):
    ECRRepository(stack, "TestECRRepository", name="test-repository", scan_on_push=False)
    assert assertions.Template.from_stack(stack).to_json() == snapshot


def test_ecr_output_disabled(stack, snapshot):
    ECRRepository(stack, "TestECRRepository", name="test-repository", outputs_enabled=False)
    assert assertions.Template.from_stack(stack).to_json() == snapshot
