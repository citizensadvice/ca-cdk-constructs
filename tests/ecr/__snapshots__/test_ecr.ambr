# name: test_ecr_additional_accounts_pull
  dict({
    'Outputs': dict({
      'RepositoryArn': dict({
        'Description': 'ARN for ECR Repository test-repository',
        'Value': dict({
          'Fn::GetAtt': list([
            'TestECRRepository654024A6',
            'Arn',
          ]),
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': True,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
          'RepositoryPolicyText': dict({
            'Statement': list([
              dict({
                'Action': list([
                  'ecr:BatchCheckLayerAvailability',
                  'ecr:GetDownloadUrlForLayer',
                  'ecr:BatchGetImage',
                ]),
                'Effect': 'Allow',
                'Principal': dict({
                  'AWS': dict({
                    'Fn::Join': list([
                      '',
                      list([
                        'arn:',
                        dict({
                          'Ref': 'AWS::Partition',
                        }),
                        ':iam::123456789012:root',
                      ]),
                    ]),
                  }),
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_ecr_additional_accounts_push
  dict({
    'Outputs': dict({
      'RepositoryArn': dict({
        'Description': 'ARN for ECR Repository test-repository',
        'Value': dict({
          'Fn::GetAtt': list([
            'TestECRRepository654024A6',
            'Arn',
          ]),
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': True,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
          'RepositoryPolicyText': dict({
            'Statement': list([
              dict({
                'Action': list([
                  'ecr:CompleteLayerUpload',
                  'ecr:UploadLayerPart',
                  'ecr:InitiateLayerUpload',
                  'ecr:BatchCheckLayerAvailability',
                  'ecr:PutImage',
                ]),
                'Effect': 'Allow',
                'Principal': dict({
                  'AWS': dict({
                    'Fn::Join': list([
                      '',
                      list([
                        'arn:',
                        dict({
                          'Ref': 'AWS::Partition',
                        }),
                        ':iam::123456789012:root',
                      ]),
                    ]),
                  }),
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_ecr_additional_rules
  dict({
    'Outputs': dict({
      'RepositoryArn': dict({
        'Description': 'ARN for ECR Repository test-repository',
        'Value': dict({
          'Fn::GetAtt': list([
            'TestECRRepository654024A6',
            'Arn',
          ]),
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': True,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":10,"description":"Test rule","selection":{"tagStatus":"untagged","countType":"imageCountMoreThan","countNumber":10},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_ecr_defaults
  dict({
    'Outputs': dict({
      'RepositoryArn': dict({
        'Description': 'ARN for ECR Repository test-repository',
        'Value': dict({
          'Fn::GetAtt': list([
            'TestECRRepository654024A6',
            'Arn',
          ]),
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': True,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_ecr_output_disabled
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': True,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_ecr_scan_disabled
  dict({
    'Outputs': dict({
      'RepositoryArn': dict({
        'Description': 'ARN for ECR Repository test-repository',
        'Value': dict({
          'Fn::GetAtt': list([
            'TestECRRepository654024A6',
            'Arn',
          ]),
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'TestECRRepository654024A6': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'ImageScanningConfiguration': dict({
            'ScanOnPush': False,
          }),
          'LifecyclePolicy': dict({
            'LifecyclePolicyText': '{"rules":[{"rulePriority":19,"description":"Delete untagged images","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countNumber":1,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":18,"description":"Delete development images after 90 days year","selection":{"tagStatus":"tagged","tagPrefixList":["dev_"],"countType":"sinceImagePushed","countNumber":90,"countUnit":"days"},"action":{"type":"expire"}},{"rulePriority":20,"description":"Limit max number of images","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1000},"action":{"type":"expire"}}]}',
          }),
          'RepositoryName': 'test-repository',
        }),
        'Type': 'AWS::ECR::Repository',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
